### ISSUE TITLE
Poor Error for function implicitly returning `None`

### ISSUE THREAD

**Issue by aDotInTheVoid** [microsoft/pyright#4466]
**Describe the bug**

When run over a function that implicitly returns `None` due to not all if cases handled, error points to function signature, not missing location in function body.

**To Reproduce**

```python
def get_int(b: bool) -> int:
    if b:
        return 1
```

Outputs:

```
/home/nixon/dev/tycheshow-py/bad_err.py
  /home/nixon/dev/tycheshow-py/bad_err.py:1:25 - error: Function with declared type of "int" must return value
    Type "None" cannot be assigned to type "int" (reportGeneralTypeIssues)
```

**Expected behavior**

The error message should explain why `None` may be returned. Ideally it will mention that the `if` statement doesn't have an else, and that this implicitly returns `None`

**VS Code extension or command-line**

`pyright 1.1.289` from command line

**Additional context**

<details>

<summary> Full Output </summary>

```
No configuration file found.
pyproject.toml file found at /home/nixon/dev/tycheshow-py.
Loading pyproject.toml file at /home/nixon/dev/tycheshow-py/pyproject.toml
Assuming Python version 3.10
Assuming Python platform Linux
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding **/.*
stubPath /home/nixon/dev/tycheshow-py/typings is not a valid directory.
Searching for source files
Found 1 source file
pyright 1.1.289
/home/nixon/dev/tycheshow-py/bad_err.py
  /home/nixon/dev/tycheshow-py/bad_err.py:1:25 - error: Function with declared type of "int" must return value
    Type "None" cannot be assigned to type "int" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
Completed in 0.576sec
```

</details>

**Comment by erictraut** [comment]
I think the error message is sufficiently clear as is.

**Comment by aDotInTheVoid** [comment]
Sure, for such a small function it's fine, but in a larger case it's quite unhelpful.

Consider

```python
def to_int(x: str, some_other_cond: bool) -> int:
    match x:
        case "q":
            return 12
        case "w":
            return 43
        case "e":
            return 235
        case "r":
            return 2346
        case "t":
            return 2436
        case "y":
            return 23
        case "u":
            return 324
        case "i":
            return 6423
        case "o":
            return 234
        case "p":
            if some_other_cond:
                return 3246
            else:
                return 4321
        case "a":
            return 3246
        case "s":
            return 234
        case "d":
            return 324
        case "f":
            return 24
        case "g":
            if some_other_cond:
                return 4532
            else:
                return -123
        case "h":
            return 3245
        case "j":
            return 5432
        case "k":
            return 3245
        case "l":
            if some_other_cond:
                return 6432
        case "z":
            return 63
        case "x":
            return 4632
        case "c":
            return 324
        case "v":
            return 2346
        case "b":
            return 6432
        case "n":
            return 4632
        case "m":
            return 5432
        case _:
            return 1
```

Pyright reports the issue as being on the first line, with no other errors.

```
/home/nixon/dev/tycheshow-py/bad_error.py
  /home/nixon/dev/tycheshow-py/bad_error.py:1:46 - error: Function with declared type of "int" must return value
    Type "None" cannot be assigned to type "int" (reportGeneralTypeIssues)
``` 

![image](https://user-images.githubusercontent.com/28781354/212566058-96c52272-945c-4c4b-acca-eac2fcdd2c1f.png)


This forces the user to manually go through all the cases to work out which of them is returning `None`.

A much more usefull thing to do would be to point to the code around line 45 (the `case "l":` )and mention it as this is the source of the `None`, because currently all the error message says is that their is an error of converting `None` to `int` somewhere in this function body, but with no indication as to where this conversion occurs, I don't think it's sufficiently clear.

**Comment by erictraut** [comment]
I still think the existing error message is quite clear already, but I've added the words "on all code paths" to the end of it to help eliminate any confusion. This will be included in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.290, which I just published. It will also be included in a future release of pylance.
